src/smt/theory_recfun.cpp

```diff
@@ -0,0 +1,363 @@
+
+#include "util/stats.h"
+#include "ast/ast_util.h"
+#include "smt/theory_recfun.h"
+#include "smt/params/smt_params_helper.hpp"
+
+#define DEBUG(x) \
+    do { \
+    TRACE("recfun", tout << x << '\n';); \
+    auto& out = std::cout; out << "recfun: "; out << x; out << '\n' << std::flush; } while(0)
+
+// NOTE: debug
+struct pp_lits {
+    smt::context & ctx;
+    smt::literal *lits;
+    unsigned len;
+    pp_lits(smt::context & ctx, unsigned len, smt::literal *lits) : ctx(ctx), lits(lits), len(len) {}
+};
+
+std::ostream & operator<<(std::ostream & out, pp_lits const & pp) {
+    out << "clause{";
+    bool first = true;
+    for (auto i = 0; i < pp.len; ++i) {
+        if (first) { first = false; } else { out << " ∨ "; }
+        pp.ctx.display_detailed_literal(out, pp.lits[i]);
+    }
+    return out << "}";
+}
+
+namespace smt {
+
+    theory_recfun::theory_recfun(ast_manager & m)
+        : theory(m.mk_family_id("recfun")), m_util(0), m_trail(*this),
+          m_guards(), m_max_depth(0), m_q_case_expand(), m_q_body_expand()
+        {
+            recfun_decl_plugin * plugin =
+                reinterpret_cast<recfun_decl_plugin*>(m.get_plugin(get_family_id()));
+            SASSERT(plugin);
+            m_util = & plugin->u();
+            SASSERT(m_util);
+        }
+
+    theory_recfun::~theory_recfun() {}
+
+    char const * theory_recfun::get_name() const { return "recfun"; }
+
+    void theory_recfun::setup_params() {
+        // obtain max depth via parameters
+        smt_params_helper p(get_context().get_params());
+        set_max_depth(p.recfun_max_depth());
+    }
+
+    theory* theory_recfun::mk_fresh(context* new_ctx) {
+        return alloc(theory_recfun, new_ctx->get_manager());
+    }
+
+    bool theory_recfun::internalize_atom(app * atom, bool gate_ctx) {
+        context & ctx = get_context();
+        if (! ctx.b_internalized(atom)) {
+            bool_var v = ctx.mk_bool_var(atom);
```

Comment:
You probably want to attach the Boolean variable with the theory.
See for example theory_array.cpp line 254.

Reply:
It doesn't seem to be sufficient, still no calls to `assign_eh` 😕

Reply:
Ah, it seems that now, marking the boolean variable as relevant does the trick. I'll dive into debugging again, thank you!!